### PR TITLE
Implements replace unknown target tokens by the source token with the highest attention

### PR DESCRIPTION
**Pull request by jordimas**
Implements replace unknown target tokens by the source token with the highest attention

This mimics OpenNMT-tf which provides this functionality when doing inference

### CODE REVIEW

File: src/translator.cc [hunk 1]
```diff
@@ -129,6 +129,24 @@ namespace ctranslate2 {
     return results;
   }
 
+  void
```
Comment (guillaumekln) [inline src/translator.cc:132]: It looks like this function does not need to be a member of `Translator`. Can you make it a static function instead and remove the declaration in `translator.h`?

```cpp
static void replace_unknowns(...)
```

File: src/translator.cc [hunk 2]
```diff
@@ -129,6 +129,24 @@ namespace ctranslate2 {
...
+  Translator::replace_unknowns(const std::vector<std::string>& source,
+                               std::vector<std::vector<std::string>>& hypotheses,
+                               const std::vector<std::vector<std::vector<float>>>& attention) {
+      for (size_t h = 0; h < hypotheses.size(); ++h) {
+        for (size_t t = 0; t < hypotheses[h].size(); ++t) {
+          if (hypotheses[h][t] == Vocabulary::unk_token) {
+            const std::vector<float> attention_values = attention[h][t];
```
Comment (guillaumekln) [inline src/translator.cc:139]: Avoid a copy:

```suggestion
            const std::vector<float>& attention_values = attention[h][t];
```

File: src/translator.cc [hunk 3]
```diff
@@ -129,6 +129,24 @@ namespace ctranslate2 {
...
+            int pos = std::distance(attention_values.begin(),std::max_element(attention_values.begin(),
+                                attention_values.end()));
```
Comment (guillaumekln) [inline src/translator.cc:141]: Improve the formatting:

```suggestion
            const size pos = std::distance(attention_values.begin(),
                                           std::max_element(attention_values.begin(), attention_values.end()));
```

File: src/translator.cc [hunk 4]
```diff
@@ -129,6 +129,24 @@ namespace ctranslate2 {
...
+
+            std::string source_token = source[pos];
+            hypotheses[h][t] = source_token;
```
Comment (guillaumekln) [inline src/translator.cc:144]: Avoid the intermediate copy:

```suggestion
            hypotheses[h][t] = source[pos];
```

File: src/translator.cc [hunk 5]
```diff
@@ -203,13 +221,24 @@ namespace ctranslate2 {
...
 
     // Convert generated ids to tokens.
     std::vector<TranslationResult> final_results;
     final_results.reserve(results.size());
     for (size_t i = 0; i < batch_size; ++i) {
       GenerationResult<size_t>& result = results[i];
+      std::vector<std::vector<std::string>> hypotheses = target_vocabulary.to_tokens(result.hypotheses());
+  
+      if (result.has_attention() && options.replace_unknowns) {
+          auto attention_values = results[i].attention();
```
Comment (guillaumekln) [inline src/translator.cc:234]: Avoid a copy:

```suggestion
          const auto& attention_values = results[i].attention();
```

**Review by guillaumekln — CHANGES_REQUESTED** [review]
Thanks! I'm OK to include this since the required changes are small.

Can you also add the new option to the following entrypoints?

* In the `translate_file` Python method
* In the client `cli/translate.cc`

File: src/translator.cc (hunk 1 above)
Comment (jordimas) [reply src/translator.cc]: Sure.

File: src/translator.cc [hunk 6]
```diff
@@ -129,6 +129,23 @@ namespace ctranslate2 {
     return results;
   }
 
+  static void
+  replace_unknowns(const std::vector<std::string>& source,
+                               std::vector<std::vector<std::string>>& hypotheses,
+                               const std::vector<std::vector<std::vector<float>>>& attention) {
```
Comment (guillaumekln) [inline src/translator.cc:135]: Can you update the indentation here?

File: src/translator.cc [hunk 7]
```diff
@@ -203,13 +220,24 @@ namespace ctranslate2 {
...
+      std::vector<std::vector<std::string>> hypotheses = target_vocabulary.to_tokens(result.hypotheses());
+  
+      if (result.has_attention() && options.replace_unknowns) {
+          const auto& attention_values = results[i].attention();
+          replace_unknowns(source[i], hypotheses, attention_values);
+
+          if (!options.return_attention) {
+            std::vector<std::vector<std::vector<float>>> empty_attention;
+            result.set_attention(empty_attention);
+          }
```
Comment (guillaumekln) [inline src/translator.cc:239]: Can you use 2-space indentation here?

**Review by guillaumekln — COMMENTED** [review]
Thanks! Just some additional comments regarding indentation:

**Comment by jordimas** [comment]
Thanks! Great feedback. I implemented all the changes and tested all the modules (translate_file, cli, etc). 

Let me know if any additional change is needed

File: src/translator.cc (hunk 6 above)
Comment (jordimas) [reply src/translator.cc]: Sure, done.

File: src/translator.cc (hunk 7 above)
Comment (jordimas) [reply src/translator.cc]: Sure. Done.

**Review by guillaumekln — APPROVED** [review]

**Pull request merged**